### PR TITLE
chore: add side-by-side tests with official proto.Marshal and Unmarshal

### LINKED ISSUE #2: let's add side-by-side with "official" protobuf marshaling
> let's add side-by-side with "official" protobuf marshaling
> _Originally posted by @smira in https://github.com/siderolabs/protoenc/pull/1#pullrequestreview-1048025699_

## Task

We should ensure that our encoder/decoder works with binary data and structs produced by official protobuf compiler. For that each test should basically contain the following.

1. Marshal Go structs to binary form using our own Marshaller
2. Unmarshal binary form into generated Go structs compiled by `protobuf` from proto messages using `proto.Unmarshal`
3. Ensure that fields in two structs match.
4. Encode resulting generated Go structs back to binary form using `proto.Marshal`
5. Unmarshal binary form into original Go structs using our own Unmarshaller.
6. Ensure that resulting struct match the original one from the stage 1.

The following test cases should be implemented:
- [x] Messages with simple fields.
- [x] Messages with repeated simple fields.
- [x] Messages with maps of simple fields.
- [x] Messages containing a map where the key is a simple type and the value is a message that contains variations of the above.
- [x] Complex messages with inner messages which contain either maps or repeated fields or both in various forms.

`anyof` and to lesser extent `enum` are not supported for now.

## Questions

- [x] Does kres support code generation for protobuf messages that are not attached to GRPC methods?

### PR DESCRIPTION
**Pull request by DmitriyMV**
This tests ensures that our encoding is compatible with official proto modules.

Closes #2

Signed-off-by: Dmitriy Matrenichev <dmitry.matrenichev@siderolabs.com>

### CODE REVIEW

File: messages/messages_test.go [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+type BasicMessage struct {
+	Int64      int64             `protobuf:"1"`
+	UInt64     uint64            `protobuf:"3"`
+	Fixed64    protoenc.FixedU64 `protobuf:"5"`
+	SomeString string            `protobuf:"6"`
+	SomeBytes  []byte            `protobuf:"7"`
+}
+
+func TestBasicMessage(t *testing.T) {
+	t.Run("check that output of both messages is the same", func(t *testing.T) {
```
Comment (sauterp) [inline messages/messages_test.go:27]:
```suggestion
	t.Run("check that the outputs of both messages are the same", func(t *testing.T) {
```

File: messages/messages_test.go [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+
+		encoded1 := must(protoenc.Marshal(&ourBasicMessage))(t)
+		basicMessage := protoUnmarshal[messages.BasicMessage](t, encoded1)
+		encoded2 := must(proto.Marshal(basicMessage))(t)
+		decodedMessage := ourUnmarshal[BasicMessage](t, encoded2)
+
+		shouldBeEqual(t, ourBasicMessage, decodedMessage)
+	})
+
+	t.Run("check that output of both zero messages is the same", func(t *testing.T) {
```
Comment (sauterp) [inline messages/messages_test.go:44]:
```suggestion
	t.Run("check that the outputs of both zero messages are the same", func(t *testing.T) {
```

File: messages/messages_test.go [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+		ourBasicMessage := BasicMessage{}
+		encoded1 := must(protoenc.Marshal(&ourBasicMessage))(t)
+		basicMessage := protoUnmarshal[messages.BasicMessage](t, encoded1)
+		encoded2 := must(proto.Marshal(basicMessage))(t)
+		decodedMessage := ourUnmarshal[BasicMessage](t, encoded2)
+
+		shouldBeEqual(t, ourBasicMessage, decodedMessage)
+	})
+
+	t.Run("check that output of both somewhat empty messages is the same", func(t *testing.T) {
```
Comment (sauterp) [inline messages/messages_test.go:54]:
```suggestion
	t.Run("check that the outputs of both somewhat empty messages are the same", func(t *testing.T) {
```

File: messages/messages_test.go [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+type MessageRepeatedFields struct {
+	Int64      []int64             `protobuf:"1"`
+	UInt64     []uint64            `protobuf:"3"`
+	Fixed64    []protoenc.FixedU64 `protobuf:"5"`
+	SomeString []string            `protobuf:"6"`
+	SomeBytes  [][]byte            `protobuf:"7"`
+}
+
+func TestMessageRepeatedFields(t *testing.T) {
+	t.Run("check that output of both messages is the same", func(t *testing.T) {
```
Comment (sauterp) [inline messages/messages_test.go:74]:
```suggestion
	t.Run("check that the outputs of both messages are the same", func(t *testing.T) {
```

File: messages/messages_test.go [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+
+		encoded1 := must(protoenc.Marshal(&originalMsg))(t)
+		decodedMsg := protoUnmarshal[messages.MessageRepeatedFields](t, encoded1)
+		encoded2 := must(proto.Marshal(decodedMsg))(t)
+		resultMsg := ourUnmarshal[MessageRepeatedFields](t, encoded2)
+
+		shouldBeEqual(t, originalMsg, resultMsg)
+	})
+
+	t.Run("check that output of both zero messages is the same", func(t *testing.T) {
```
Comment (sauterp) [inline messages/messages_test.go:91]:
```suggestion
	t.Run("check that the outputs of both zero messages are the same", func(t *testing.T) {
```

File: messages/messages_test.go [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+		ourBasicMessage := MessageRepeatedFields{}
+		encoded1 := must(protoenc.Marshal(&ourBasicMessage))(t)
+		basicMessage := protoUnmarshal[messages.MessageRepeatedFields](t, encoded1)
+		encoded2 := must(proto.Marshal(basicMessage))(t)
+		decodedMessage := ourUnmarshal[MessageRepeatedFields](t, encoded2)
+
+		shouldBeEqual(t, ourBasicMessage, decodedMessage)
+	})
+
+	t.Run("check that output of both somewhat empty messages is the same", func(t *testing.T) {
```
Comment (sauterp) [inline messages/messages_test.go:101]:
```suggestion
	t.Run("check that the outputs of both somewhat empty messages are the same", func(t *testing.T) {
```

File: messages/messages_test.go [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+		shouldBeEqual(t, ourBasicMessage, decodedMessage)
+	})
+}
+
+type BasicMessageRep struct {
+	BasicMessage []BasicMessage `protobuf:"1"`
+}
+
+func TestBasicMessageRep(t *testing.T) {
+	t.Run("check that output of both messages is the same", func(t *testing.T) {
```
Comment (sauterp) [inline messages/messages_test.go:117]:
```suggestion
	t.Run("check that the outputs of both messages are the same", func(t *testing.T) {
```

File: messages/messages_test.go [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+		}
+		encoded1 := must(protoenc.Marshal(&originalMsg))(t)
+		decodedMsg := protoUnmarshal[messages.BasicMessageRep](t, encoded1)
+		encoded2 := must(proto.Marshal(decodedMsg))(t)
+		resultMsg := ourUnmarshal[BasicMessageRep](t, encoded2)
+
+		shouldBeEqual(t, originalMsg, resultMsg)
+	})
+
+	t.Run("check that output of both zero messages is the same", func(t *testing.T) {
```
Comment (sauterp) [inline messages/messages_test.go:144]:
```suggestion
	t.Run("check that the outputs of both zero messages are the same", func(t *testing.T) {
```

File: messages/messages_test.go [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+		ourBasicMessage := BasicMessageRep{}
+		encoded1 := must(protoenc.Marshal(&ourBasicMessage))(t)
+		basicMessage := protoUnmarshal[messages.BasicMessageRep](t, encoded1)
+		encoded2 := must(proto.Marshal(basicMessage))(t)
+		decodedMessage := ourUnmarshal[BasicMessageRep](t, encoded2)
+
+		shouldBeEqual(t, ourBasicMessage, decodedMessage)
+	})
+
+	t.Run("check that output of both somewhat empty messages is the same", func(t *testing.T) {
```
Comment (sauterp) [inline messages/messages_test.go:154]:
```suggestion
	t.Run("check that the outputs of both somewhat empty messages are the same", func(t *testing.T) {
```

File: messages/messages_test.go [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+}
+
+type MessageComplexFields struct {
+	MapToMsg     map[string]*BasicMessage    `protobuf:"1"`
+	MapToMsgs    map[string]*BasicMessageRep `protobuf:"2"`
+	PrimitiveMap map[string]int64            `protobuf:"3"`
+}
+
+func TestMessageComplexFields(t *testing.T) {
+	t.Run("check that output of both messages is the same", func(t *testing.T) {
```
Comment (sauterp) [inline messages/messages_test.go:178]:
```suggestion
	t.Run("check that the outputs of both messages are the same", func(t *testing.T) {
```

File: messages/messages_test.go [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+		}
+		encoded1 := must(protoenc.Marshal(&originalMsg))(t)
+		decodedMsg := protoUnmarshal[messages.MessageComplexFields](t, encoded1)
+		encoded2 := must(proto.Marshal(decodedMsg))(t)
+		resultMsg := ourUnmarshal[MessageComplexFields](t, encoded2)
+
+		shouldBeEqual(t, originalMsg, resultMsg)
+	})
+
+	t.Run("check that output of both zero messages is the same", func(t *testing.T) {
```
Comment (sauterp) [inline messages/messages_test.go:241]:
```suggestion
	t.Run("check that the outputs of both zero messages are the same", func(t *testing.T) {
```

File: messages/messages_test.go [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+		ourBasicMessage := MessageComplexFields{}
+		encoded1 := must(protoenc.Marshal(&ourBasicMessage))(t)
+		basicMessage := protoUnmarshal[messages.MessageComplexFields](t, encoded1)
+		encoded2 := must(proto.Marshal(basicMessage))(t)
+		decodedMessage := ourUnmarshal[MessageComplexFields](t, encoded2)
+
+		shouldBeEqual(t, ourBasicMessage, decodedMessage)
+	})
+
+	t.Run("check that output of both somewhat empty messages is the same", func(t *testing.T) {
```
Comment (sauterp) [inline messages/messages_test.go:251]:
```suggestion
	t.Run("check that the outputs of both somewhat empty messages are the same", func(t *testing.T) {
```

**Review by sauterp — APPROVED** [review]

File: messages/messages.proto [hunk 13]
```diff
@@ -0,0 +1,30 @@
+syntax = "proto3";
```
Comment (smira) [inline messages/messages.proto:1]: we can make Kres build this: 

https://github.com/cosi-project/runtime/blob/master/.kres.yaml

File: messages/messages_test.go (hunk 1 above)
Comment (DmitriyMV) [reply messages/messages_test.go]: Thanks!

File: messages/messages.proto (hunk 13 above)
Comment (DmitriyMV) [reply messages/messages.proto:1]: Done, I think.

**Pull request merged**